### PR TITLE
Record controller logs in the events table

### PR DESCRIPTION
**Pull request by wesbillman**
Screenshot 2023-08-15 at 2 03 50 PM

### CODE REVIEW

File: backend/controller/controller.go [hunk 1]
```diff
@@ -98,10 +98,14 @@ type Service struct {
 	deploymentReservationTimeout time.Duration
 	artefactChunkSize            int
 	key                          model.ControllerKey
+	ctx                          context.Context
```
Comment (wesbillman) [inline backend/controller/controller.go:101]: I added this `ctx` here because the sink is processing logs in the background and the request `ctx` seems to get canceled before I get a chance to write the logs to the db. So I'm using the `controller`'s context for the sink. Maybe there's a better way?

File: backend/controller/controller.go [hunk 2]
```diff
@@ -98,10 +98,14 @@ type Service struct {
...
 
 	clientsMu sync.Mutex
 	// Map from endpoint to client.
 	clients map[string]clients
+	logger  *log.Logger
+	// Map from deployment key to logger.
+	deploymentLoggers map[model.DeploymentKey]*log.Logger
```
Comment (wesbillman) [inline backend/controller/controller.go:108]: Is this ok? Since we can have multiple deployments it seems like we need multiple loggers for them. This part felt a bit odd to me.

File: backend/controller/deployment_logs.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+// Log implements Sink
+func (d *deploymentLogsSink) Log(entry log.Entry) error {
+	var request types.Option[model.IngressRequestKey]
+	if reqStr, ok := entry.Attributes["request"]; ok {
+		req, err := model.ParseIngressRequestKey(reqStr)
+		if err == nil {
+			request = types.Some(req)
+		}
+	}
```
Comment (wesbillman) [inline backend/controller/deployment_logs.go:50]: Not sure if this is needed. Will we have `request` attributes here?

File: backend/runner/runner.go [hunk 4]
```diff
@@ -368,7 +368,7 @@ func (s *Service) streamLogsLoop(ctx context.Context, send func(request *ftlv1.S
 		err := send(&ftlv1.StreamDeploymentLogsRequest{
 			RequestKey:    requestKey,
 			DeploymentKey: deployment.key.String(),
-			TimeStamp:     entry.Time.Unix(),
+			TimeStamp:     entry.Time.UnixMilli(),
```
Comment (wesbillman) [inline backend/runner/runner.go:371]: I noticed we were dropping the milliseconds in the logs from the `runner` this seems to fix that issue.

File: backend/controller/controller.go (hunk 1 above)
Comment (alecthomas) [reply backend/controller/controller.go:101]: Yes, you can use `unstoppable.Context(ctx)`.

File: backend/controller/controller.go [hunk 5]
```diff
@@ -316,13 +328,20 @@ func (s *Service) ReplaceDeploy(ctx context.Context, c *connect.Request[ftlv1.Re
 	if err != nil {
 		return nil, connect.NewError(connect.CodeInvalidArgument, errors.WithStack(err))
 	}
+
```
Comment (alecthomas) [inline backend/controller/controller.go:331]: I think you'll want two loggers here? One for the old deployment, and one for the new one.

File: backend/runner/runner.go (hunk 4 above)
Comment (alecthomas) [reply backend/runner/runner.go:371]: We should be using proto [Time](https://protobuf.dev/reference/protobuf/google.protobuf/#timestamp), not raw integers. Same with durations. There are helper functions for converting to and from Go types in `timestamppb`.

File: backend/controller/deployment_logs.go (hunk 3 above)
Comment (alecthomas) [reply backend/controller/deployment_logs.go:50]: I concluded that we'll have to have it in the logs, because we can't get it from anywhere else as the logger doesn't accept a Context.

TL;DR I couldn't think of a better idea, and I think this makes sense as we're doing it for other keys.

File: backend/controller/controller.go [hunk 6]
```diff
@@ -929,3 +966,19 @@ func extractIngressRoutingEntries(req *ftlv1.CreateDeploymentRequest) []dal.Ingr
...
+	}
+
+	deploymentLogger, ok := s.deploymentLoggers[*deploymentKey]
+	if ok {
+		return deploymentLogger
+	}
+
+	// create deployment logger if it doesn't exist
+	logger := log.FromContext(ctx).AddSink(newDeploymentLogsSink(s.ctx, *deploymentKey, s.dal))
+	s.deploymentLoggers[*deploymentKey] = logger
```
Comment (alecthomas) [inline backend/controller/controller.go:982]: This is not thread safe. You could either use a `sync.Map`, or wrap all access (reads and writes) in a `sync.Mutex`.

Comment (alecthomas) [reply backend/controller/controller.go:982]: You'll also need some system for deleting them, which I don't see here?

File: backend/controller/internal/dal/dal.go [hunk 7]
```diff
@@ -348,13 +348,7 @@ type IngressRoutingEntry struct {
...
 //
 // If an existing deployment with identical artefacts exists, it is returned.
-func (d *DAL) CreateDeployment(
-	ctx context.Context,
-	language string,
-	schema *schema.Module,
-	artefacts []DeploymentArtefact,
-	ingressRoutes []IngressRoutingEntry,
-) (key model.DeploymentKey, err error) {
+func (d *DAL) CreateDeployment(ctx context.Context, deploymentKey model.DeploymentKey, language string, schema *schema.Module, artefacts []DeploymentArtefact, ingressRoutes []IngressRoutingEntry) (key model.DeploymentKey, err error) {
```
Comment (alecthomas) [inline backend/controller/internal/dal/dal.go:351]: No point returning the key that you've passed in.

**Comment by alecthomas** [comment]
I'll merge this in but can you send a followup addressing the remaining couple of comments?
